### PR TITLE
🎉 Source Greenhouse: Implement Additional Endpoints Support

### LINKED ISSUE #1943: Adopt connector best practices for source greenhouse
## Tell us about the problem you're trying to solve
We'd like to follow [the published connector development best practices ](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector/best-practices-for-connector-development) on this connector. 

Since this connector is Singer based, for each of the below items please verify whether:
- the singer tap already implements it
- the singer tap does not implement it and would require forking the tap to implement
- it's something we can implement easily on top of the Singer tap without forking it

**If implementing any of the below items would require forking the Singer tap, please notify @sherifnada and move on to the next issue.** 

## Checklist
- [ ] Check connection verifies permissions upfront
- [ ] Check connection failure emits actionable error messages. This will probably require you to think about the failure modes that can occur (e.g: timeout, permission not granted, token does not exist, etc.. -- make sure that you add custom integration tests to verify each of these failure modes, or unit tests if you can mock the HTTP requests involved)
- [ ] If API connector: integration tests validate that every stream outputs data (you may need to seed the underlying API with fake data). This will establish a baseline confidence that the connector works correctly on every data stream. If this is not possible, please reach out to @sherifnada with the reason to discuss how to move forward. 
- [ ] If API connector: backs off on hitting rate limits (instead of just failing/exiting)
- [ ] If API connector: implements incremental sync (where possible -- not all streams will allow this)

### LINKED ISSUE #5345: Greenhouse: Support a few more endpoints
## Tell us about the problem you're trying to solve
We're looking to measure a few diversity questions from Greenhouse and I believe this is tracked in an endpoint that is not currently available from the Greenhouse connector, which is the endpoint **eeoc**
There are a few more that would be useful:

- demographic data
- scheduled interviews
- sources
- rejection reasons
- job openings
- job stage

## Describe the solution you’d like
More output streams from Greenhouse connector

## Describe the alternative you’ve considered or used
N/A

## Are you willing to submit a PR?
Not technical enough!

### PR DESCRIPTION
**Pull request by htrueman**
## What
Closes #5345

## How
Implement nested stream support.
Add new schemas.
Update acceptance test config.

## Recommended reading order
1. `client.py`
2. Schemas
3. `acceptance-test-config.yml`

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector added to connector index like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

**Comment by htrueman** [comment]
Screenshot 2021-09-24 at 09 33 30

For some reason acceptance tests fail with 429 error in github. I'm trying to figure out the reason, but please do a review.

File: airbyte-integrations/connectors/source-greenhouse/build.gradle [hunk 1]
```diff
@@ -1,21 +1,14 @@
...
-    configuredCatalogPath = "sample_files/configured_catalog.json"
-}
-
-
 dependencies {
-    implementation files(project(':airbyte-integrations:bases:base-standard-source-test-file').airbyteDocker.outputs)
+    implementation files(project(':airbyte-integrations:bases:source-acceptance-test').airbyteDocker.outputs)
     implementation files(project(':airbyte-integrations:bases:base-python').airbyteDocker.outputs)
-}
+}
```
Comment (Zirochkaa) [inline airbyte-integrations/connectors/source-greenhouse/build.gradle:14]: Please add an empty line at the end of the file.

File: airbyte-integrations/connectors/source-greenhouse/acceptance-test-config.yml [hunk 2]
```diff
@@ -17,6 +17,19 @@ tests:
...
+          "applications.demographics_answers",
+          "applications.interviews",
+          "demographics_answer_options",
+          "demographics_answers",
+          "demographics_answers.answer_options",
+          "demographics_question_sets",
+          "demographics_question_sets.questions",
+          "demographics_questions",
+          "interviews",
+          "rejection_reasons",
```
Comment (Zirochkaa) [inline airbyte-integrations/connectors/source-greenhouse/acceptance-test-config.yml:31]: Is there a way to fill those streams? If yes, then maybe we could create a ticket for it?

File: airbyte-integrations/connectors/source-greenhouse/source_greenhouse/client.py [hunk 3]
```diff
@@ -35,44 +35,103 @@
...
     def __init__(self, api_key):
-        self._client = Harvest(api_key=api_key)
+        self._client = HarvestClient(api_key=api_key)
         super().__init__()
 
     def list(self, name, **kwargs):
-        yield from paginator(getattr(self._client, name), **kwargs)
-
-    def _enumerate_methods(self) -> Mapping[str, callable]:
+        name_parts = name.split(".")
```
Comment (Zirochkaa) [inline airbyte-integrations/connectors/source-greenhouse/source_greenhouse/client.py:116]: Could you please add an annotation to `list` function?

**Review by Zirochkaa — APPROVED** [review]

**Comment by Zirochkaa** [comment]
Also could you please update `greenhouse.md` file and put links to docs for all newly added endpoints there?

File: airbyte-integrations/connectors/source-greenhouse/source_greenhouse/client.py [hunk 4]
```diff
@@ -35,44 +35,103 @@
...
+        for row in rows:
+            yield from paginator(getattr(request(**params, object_id=row["id"]), nested_names[0]))
+    else:
+        yield from rows
     while request.records_remaining:
         rows = request.get_next()
         yield from rows
 
 
+class HarvestClient(Harvest):
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-greenhouse/source_greenhouse/client.py:56]: we should make a ticket to convert this to the CDK

File: airbyte-integrations/connectors/source-greenhouse/acceptance-test-config.yml (hunk 2 above)
Comment (sherifnada) [reply airbyte-integrations/connectors/source-greenhouse/acceptance-test-config.yml:31]: ^

**Review by sherifnada — COMMENTED** [review]
Can we fill this with data? as it stands we're not actually testing many of these streams very well

File: airbyte-integrations/connectors/source-greenhouse/source_greenhouse/client.py (hunk 4 above)
Comment (htrueman) [reply airbyte-integrations/connectors/source-greenhouse/source_greenhouse/client.py:56]: We have an issue for that: https://github.com/airbytehq/airbyte/issues/1943. There is also an issue to add the incremental sync support: https://github.com/airbytehq/airbyte/issues/1386. Those 2 issue may be merged into one.

I'll add a comment in the code following that.

**Comment by Zirochkaa** [comment]
image

**Review by sherifnada — CHANGES_REQUESTED** [review]
How did you test these changes? I'm asking because if these streams are marked empty in SAT, how do we know that the current e.g: schemas or new endpoints work? We should not merge in the current state because we have no guarantee this is working as far as I can tell from looking at the code

**Comment by htrueman** [comment]
@sherifnada please make a review

**Review by sherifnada — COMMENTED** [review]
Looks great! a few requests: 
1. unit tests for the base functionality
2. ideally we keep the naming `_` as opposed to `X.Y_Z` for stream names
3. because this PR is also adding new streams we should populate those streams with data before merging so we have confidence in the implementation

**Review by sherifnada — APPROVED** [review]
Looks great! Only comment is around whether to expose new streams which we can't test

**Pull request merged**